assets/src/edit-story/components/library/panes/text/textSets/textSets.js

```diff
@@ -32,34 +32,77 @@ import { __ } from '@wordpress/i18n';
 import { Section } from '../../../common';
 import { UnitsProvider } from '../../../../../units';
 import { PAGE_RATIO, TEXT_SET_SIZE } from '../../../../../constants';
+import PillGroup from '../../shared/pillGroup';
+import { PANE_PADDING, useRovingTabIndex } from '../../shared';
 import { getTextSets } from './utils';
 import TextSet from './textSet';
 
 const TextSetContainer = styled.div`
   display: grid;
   grid-template-columns: 1fr 1fr;
   row-gap: 12px;
+  column-gap: 12px;
+  margin-top: 28px;
 `;
 
+/* Undo the -1.5em set by the Pane */
+const CategoryWrapper = styled.div`
+  margin-left: -${PANE_PADDING};
+  margin-right: -${PANE_PADDING};
+`;
+
+const CATEGORIES = {
+  contact: __('Contact', 'web-stories'),
+  editorial: __('Editorial', 'web-stories'),
+  list: __('List', 'web-stories'),
+  cover: __('Cover', 'web-stories'),
+  section_header: __('Header', 'web-stories'),
+  step: __('Steps', 'web-stories'),
+  table: __('Table', 'web-stories'),
+  quote: __('Quote', 'web-stories'),
+};
+
 function TextSets() {
   const [textSets, setTextSets] = useState([]);
+  const [selectedCat, setSelectedCat] = useState(null);
+  const ref = useRef();
+
+  const getFilteredTextSets = useCallback(() => {
+    if (selectedCat && textSets[selectedCat]) {
+      return textSets[selectedCat];
+    }
+    return Object.values(textSets).flat();
+  }, [selectedCat, textSets]);
```

Comment:
Why is this a function and not just a memoized list? Seems to be called only once and on every render? Also, I don't see how `selectedCat` can be non-`null` while `textSets[selectedCat]` is `null`. And `allTextSets` should be memoized too - no need to do that whole enumeration every time the category is unselected.

```js
  const allTextSets = useMemo(() => Object.values(textSets).flat(), [textSets]);
  const filteredTextSets = useMemo(
    () => selectedCat ? textSets[selectedCat] : allTextSets,
    [selectedCat, textSets, allTextSets],
  );
```

Reply:
Because I forgot about memoizing 😞 Will add, thank you!